app/src/main/java/org/mozilla/focus/webview/SystemWebView.kt

```diff
@@ -0,0 +1,330 @@
+        }
+    }
+
+    override fun setCallback(callback: IWebView.Callback?) {
+        this.callback = callback
+        client.setCallback(callback)
+        linkHandler.setCallback(callback)
+    }
+
+    override fun setFindListener(findListener: IFindListener) {
+        this.setFindListener(findListener as WebView.FindListener)
+    }
+
+    @Suppress("deprecation")
+    override fun loadUrl(url: String?) {
+        // We need to check external URL handling here - shouldOverrideUrlLoading() is only
+        // called by webview when clicking on a link, and not when opening a new page for the
+        // first time using loadUrl().
+        if (!client.shouldOverrideUrlLoading(this, url)) {
+            val additionalHeaders = HashMap<String, String>()
+            additionalHeaders["X-Requested-With"] = ""
+
+            super.loadUrl(url, additionalHeaders)
+        }
+
+        client.notifyCurrentURL(url)
+    }
+
+    override fun exitFullscreen() {}
+
+    override fun loadData(
+        baseURL: String,
+        data: String,
+        mimeType: String,
+        encoding: String,
+        historyURL: String
+    ) {
+        loadDataWithBaseURL(baseURL, data, mimeType, encoding, historyURL)
+    }
+
+    override fun destroy() {
+        super.destroy()
+
+        // WebView might save data to disk once it gets destroyed. In this case our cleanup call
+        // might not have been able to see this data. Let's do it again.
+        deleteContentFromKnownLocations(context)
+    }
+
+    override fun cleanup() {
+        clearFormData()
+        clearHistory()
+        clearMatches()
+        clearSslPreferences()
+        clearCache(true)
+
+        // We don't care about the callback - we just want to make sure cookies are gone
+        CookieManager.getInstance().removeAllCookies(null)
+
+        WebStorage.getInstance().deleteAllData()
+
+        val webViewDatabase = WebViewDatabase.getInstance(context)
+        webViewDatabase.clearHttpAuthUsernamePassword()
+
+        deleteContentFromKnownLocations(context)
+    }
+
+    override fun autofill(values: SparseArray<AutofillValue>) {
+        super.autofill(values)
+
+        TelemetryWrapper.autofillPerformedEvent()
+    }
+
+    private fun createWebChromeClient(): WebChromeClient {
+        return object : WebChromeClient() {
+            override fun onProgressChanged(view: WebView, newProgress: Int) {
+                if (callback != null) {
+                    // This is the earliest point where we might be able to confirm a redirected
+                    // URL: we don't necessarily get a shouldInterceptRequest() after a redirect,
+                    // so we can only check the updated url in onProgressChanges(), or in onPageFinished()
+                    // (which is even later).
+                    val viewURL = view.url
+                    if (!UrlUtils.isInternalErrorURL(viewURL) && viewURL != null) {
+                        callback!!.onURLChanged(viewURL)
```

Comment:
Do we still need the `!!` if we're inside a null check?

We could also potentially flatten this if we use the null check to early return.

Reply:
About the !! inside the null check. This is something we'll see a lot, no doubt. Because callback is a var, Kotlin wisely surmises that code it doesn't know about might change its value to null from another thread. The simplest solution is to use callback?.let {} so the value inside that let block is no longer mutable.